### ISSUE TITLE
Runtime hits unknown perfomance cliff in the estimator

### ISSUE THREAD

**Issue by matklad** [near/nearcore#4771]
See 

[https://github.com/matklad/nearcore/commit/80b44a877a1d2de76b4adcf63567984d615e7c36](https://github.com/matklad/nearcore/commit/80b44a877a1d2de76b4adcf63567984d615e7c36)

In that commit, I add a benchmark which sends 150 blocks, where each
 block contains 100 deploy code transaction.

EDIT: [https://github.com/near/nearcore/issues/4771#issuecomment-933347496](https://github.com/near/nearcore/issues/4771#issuecomment-933347496) is modernized version.

The benchmark is implemented in two ways:

*   on top of parameter estimator testbed
*   on top of runtime tester

In the first case, I observe a very stage behavior, where at some point
 block processsing times severely degrades.

To reproduce, checkout this commit and run

```console
 $ cargo test --release --package runtime-params-estimator --lib --features required – v2::fast --exact --nocapture
 $ cargo test --release --package runtime-params-estimator --lib --features required – v2::slow --exact --nocapture
 ```

Look for `Runtime::apply` lines in the output, they they tell you how
 long it takes to process one apply.

In the `fast` case, every apply takes about 16ms for me. In the `slow`
 case, apply start at about 30ms, but then abruptly go to 600ms.

Would be great to understand why the time suddenly balloons!

**Comment by mina86** [comment]
Quick look indicates that the slowdown comes from the storage.  I’ve
added some more tracing and here’s output with a fast
`process_transaction` invocation followed by a slow one (they are
within the same `apply` call):

    147.38µs         Runtime::process_transaction
      145.70µs         verify_and_charge_transaction
        48.12µs          validate_transaction
        46.70µs          get_account
          46.15µs          TrieUpdate::get
        49.21µs          get_access_key
          48.65µs          TrieUpdate::get
    …
    2.64ms           Runtime::process_transaction
      2.64ms           verify_and_charge_transaction
        49.03µs          validate_transaction
        1.30ms           get_account
          1.30ms           TrieUpdate::get
        1.29ms           get_access_key
          1.29ms           TrieUpdate::get

`TrieUpdate::get` is suddenly 25 times slower and it’s called quite
a bit so it adds up.

**Comment by bowenwang1996** [comment]
@mina86 is it because the value is not cached and you have to do a trie traversal?

**Comment by mina86** [comment]
It doesn’t look like cache issues.  I haven’t measured cache
hit ratio or counted how often we hit the storage, but just by looking
at logs the numbers seem similar for fast and slow cases.  On the
other hand, `Store::get` becomes ten times slower all of a sudden:

          74.73µs          Trie::Update::get
            68.40µs          Trie::lookup
              7.45µs           Store::get
              3.50µs           Store::get
              3.16µs           Store::get
              2.10µs           Store::get
              2.82µs           Store::get
              2.79µs           Store::get
              3.38µs           Store::get
              4.09µs           Store::get
              3.77µs           Store::get
              3.07µs           Store::get
              2.45µs           Store::get
              3.23µs           Store::get
              2.91µs           Store::get
              2.31µs           Store::get
              2.42µs           Store::get
              2.63µs           Store::get
              2.59µs           Store::get
            4.83µs           Store::get

vs.

          1.31ms           Trie::Update::get
            1.24ms           Trie::lookup
              86.48µs          Store::get
              72.27µs          Store::get
              70.31µs          Store::get
              69.17µs          Store::get
              68.85µs          Store::get
              68.40µs          Store::get
              68.75µs          Store::get
              69.13µs          Store::get
              69.77µs          Store::get
              68.91µs          Store::get
              68.82µs          Store::get
              69.21µs          Store::get
              68.84µs          Store::get
              78.35µs          Store::get
              75.67µs          Store::get
              76.41µs          Store::get
              75.47µs          Store::get
            69.45µs          Store::get

So I’m now looking at that.

**Comment by mina86** [comment]
The difference is caused by how signer of the transaction is chosen.
In the fast test signers are chosen in sequence from all the available
accounts.  In the slow tests, signers are chosen at random from unused
accounts.  Making the latter use sequential accounts fixes the issue:

```
--- a/runtime/runtime-params-estimator/src/v2/support.rs
+++ b/runtime/runtime-params-estimator/src/v2/support.rs
@@ -202,12 +202,12 @@ impl<'a, 'c> TransactionBuilder<'a, 'c> {
         self.testbed.accounts[account_index].clone()
     }
     pub(crate) fn random_unused_account(&mut self) -> AccountId {
-        loop {
-            let account = self.random_account();
+        for account in &self.testbed.accounts {
             if self.testbed.used_accounts.insert(account.clone()) {
-                return account;
+                return account.clone();
             }
         }
+        panic!("Not enough accounts");
     }
     pub(crate) fn random_account_pair(&mut self) -> (AccountId, AccountId) {
         let first = self.random_account();
```
I didn’t look into how exactly this affects the database to cause the slowdown.

**Comment by mina86** [comment]
Perhaps I’ve spoken too soon. The above is correct and changing
`random_unused_account` implementation does speed up the slow test but
the opposite does not slow down the fast test:

```
--- a/runtime/runtime-params-estimator/src/v2.rs
+++ b/runtime/runtime-params-estimator/src/v2.rs
@@ -449,10 +449,20 @@ fn fast() {
         use_in_memory_store: false,
     };

-    let mut i = 0;
+    let mut used = std::collections::HashSet::new();
+    let mut rng = rand::thread_rng();
+
     for h in 0..n_blocks {
         let mut block = BlockConfig::at_height((h + 1) as u64);
         for _ in 0..block_size {
+
+            let i = loop {
+                let i = rng.gen_range(0, num_accounts as u32);
+                if used.insert(i) {
+                    break i as usize;
+                }
+            };
+
             let signer_id = accounts[i].clone();
             let receiver_id = signer_id.clone();
             let signer =
@@ -464,7 +474,6 @@ fn fast() {
                 signer,
                 actions: vec![Action::DeployContract(DeployContractAction { code: code.clone() })],
             });
-            i += 1;
         }
         s.blocks.push(block)
     }
```

It looks like it’s something else calling `random_unused_account` that
is the culprit.

**Comment by matklad** [comment]
Confirmed that the order of accounts matters here. Iterating accounts by one up or down finished fast. Iterting accounts by eleven shows the same slowdown

<detail>
<summary>diff</summary>

```diff
diff --git a/runtime/runtime-params-estimator/src/v2.rs b/runtime/runtime-params-estimator/src/v2.rs
index 6a844fcd9..597da1c01 100644
--- a/runtime/runtime-params-estimator/src/v2.rs
+++ b/runtime/runtime-params-estimator/src/v2.rs
@@ -27,7 +27,7 @@ static ALL_COSTS: &[(Cost, fn(&mut Ctx) -> GasCost)] = &[
     (Cost::ActionReceiptCreation, action_receipt_creation),
     (Cost::ActionSirReceiptCreation, action_sir_receipt_creation),
     (Cost::ActionTransfer, action_transfer),
-    (Cost::ActionCreateAccount, action_create_account),
+    // (Cost::ActionCreateAccount, action_create_account),
     (Cost::ActionDeleteAccount, action_delete_account),
     (Cost::ActionAddFullAccessKey, action_add_full_access_key),
     (Cost::ActionAddFunctionAccessKeyBase, action_add_function_access_key_base),
@@ -118,28 +118,28 @@ fn action_transfer(ctx: &mut Ctx) -> GasCost {
     total_cost - base_cost
 }
 
-fn action_create_account(ctx: &mut Ctx) -> GasCost {
-    let total_cost = {
-        let mut testbed = ctx.test_bed();
+// fn action_create_account(ctx: &mut Ctx) -> GasCost {
+//     let total_cost = {
+//         let mut testbed = ctx.test_bed();
 
-        let mut make_transaction = |mut tb: TransactionBuilder<'_, '_>| -> SignedTransaction {
-            let sender = tb.random_account();
-            let new_account =
-                AccountId::try_from(format!("{}_{}", sender, tb.rng().gen::<u64>())).unwrap();
-
-            let actions = vec![
-                Action::CreateAccount(CreateAccountAction {}),
-                Action::Transfer(TransferAction { deposit: 10u128.pow(26) }),
-            ];
-            tb.transaction_from_actions(sender, new_account, actions)
-        };
-        testbed.average_transaction_cost(&mut make_transaction)
-    };
+//         let mut make_transaction = |mut tb: TransactionBuilder<'_, '_>| -> SignedTransaction {
+//             let sender = tb.random_account();
+//             let new_account =
+//                 AccountId::try_from(format!("{}_{}", sender, tb.rng().gen::<u64>())).unwrap();
 
-    let base_cost = action_receipt_creation(ctx);
+//             let actions = vec![
+//                 Action::CreateAccount(CreateAccountAction {}),
+//                 Action::Transfer(TransferAction { deposit: 10u128.pow(26) }),
+//             ];
+//             tb.transaction_from_actions(sender, new_account, actions)
+//         };
+//         testbed.average_transaction_cost(&mut make_transaction)
+//     };
 
-    total_cost - base_cost
-}
+//     let base_cost = action_receipt_creation(ctx);
+
+//     total_cost - base_cost
+// }
 
 fn action_delete_account(ctx: &mut Ctx) -> GasCost {
     let total_cost = {
diff --git a/runtime/runtime-params-estimator/src/v2/support.rs b/runtime/runtime-params-estimator/src/v2/support.rs
index 6a635eb34..9b62e6fb2 100644
--- a/runtime/runtime-params-estimator/src/v2/support.rs
+++ b/runtime/runtime-params-estimator/src/v2/support.rs
@@ -7,7 +7,7 @@ use near_crypto::{InMemorySigner, KeyType};
 use near_primitives::hash::CryptoHash;
 use near_primitives::transaction::{Action, DeployContractAction, SignedTransaction};
 use near_primitives::types::{AccountId, Gas};
-use rand::prelude::ThreadRng;
+use rand::prelude::{SliceRandom, ThreadRng};
 use rand::Rng;
 
 use crate::cases::ratio_to_gas;
@@ -50,6 +50,7 @@ impl<'c> Ctx<'c> {
             accounts: (0..self.config.active_accounts).map(get_account_id).collect(),
             nonces: HashMap::new(),
             used_accounts: HashSet::new(),
+            idx: 0,
         }
     }
 
@@ -80,6 +81,7 @@ impl<'c> Ctx<'c> {
             accounts: proto.accounts.clone(),
             nonces: proto.nonces.clone(),
             used_accounts: HashSet::new(),
+            idx: 0,
         }
     }
 
@@ -117,6 +119,7 @@ pub(crate) struct TestBed<'c> {
     accounts: Vec<AccountId>,
     nonces: HashMap<AccountId, u64>,
     used_accounts: HashSet<AccountId>,
+    idx: usize,
 }
 
 impl<'c> TestBed<'c> {
@@ -190,24 +193,26 @@ impl<'a, 'c> TransactionBuilder<'a, 'c> {
         )
     }
 
-    pub(crate) fn rng(&mut self) -> ThreadRng {
-        rand::thread_rng()
-    }
+    // pub(crate) fn rng(&mut self) -> ThreadRng {
+    //     rand::thread_rng()
+    // }
 
     pub(crate) fn account(&mut self, account_index: usize) -> AccountId {
         get_account_id(account_index)
     }
     pub(crate) fn random_account(&mut self) -> AccountId {
-        let account_index = self.rng().gen_range(0, self.testbed.accounts.len());
+        self.testbed.idx = (self.testbed.idx + 11) % self.testbed.accounts.len();
+        let account_index = self.testbed.idx;
         self.testbed.accounts[account_index].clone()
     }
     pub(crate) fn random_unused_account(&mut self) -> AccountId {
-        loop {
+        for i in 0..1000 {
             let account = self.random_account();
             if self.testbed.used_accounts.insert(account.clone()) {
                 return account;
             }
         }
+        panic!("no unused accounts")
     }
     pub(crate) fn random_account_pair(&mut self) -> (AccountId, AccountId) {
         let first = self.random_account();
```

</detail>

**Comment by matklad** [comment]
Ok, it took me just about a week to realize I can throw flamegrapsh at it :sweat_smile: 

stepping by 11 acconts:

![image](https://user-images.githubusercontent.com/1711539/132357723-0c5fa352-08db-4738-a41f-d278f6a8467a.png)

stepping by 1:

![image](https://user-images.githubusercontent.com/1711539/132357842-cdbf1a9d-1feb-4369-a9cd-308d61f4202f.png)

profilnes recoreded with `perf record -F 9999 --call-graph dwarf  target/release/deps/runtime_params_estimator-0328c9c022713977 v2::slow --exact --nocapture` and rendered with https://github.com/KDAB/hotspot

The `rocksdb::VerifyBlockChecksum` looks very suspicious! 

I'll stop at this for the time being.

**Comment by matklad** [comment]
(meta -- does anyone know of a nice way to share per profileing results? The above is me scrennshotting hotspot ui, which doesn's seem ideal)

**Comment by bowenwang1996** [comment]
> In the fast test signers are chosen in sequence from all the available
accounts. In the slow tests, signers are chosen at random from unused
accounts

@mina86 what are the available accounts? I wonder whether this affects the trie traversal in the sense that if we go over the accounts sequentially, some trie node will be cached between the reads (if not by nearcore then on the OS level the page may be cached) while accessing them randomly makes it harder to cache anything.

**Comment by mina86** [comment]
The tests create 200k accounts and all transactions are signed by one of them. The test uses each account only once. Sequential reads indeed seem like a plausible explanation though at this moment I think this is affecting caching in RocksDB or on OS level rather than in our code. (I still need to just add counters or profile each of the Trie::* and Store::* methods to be sure).

**Comment by mina86** [comment]
Turns out

```
diff --git a/genesis-tools/genesis-populate/src/lib.rs b/genesis-tools/genesis-populate/src/lib.rs
index d3a9793d..3eceddfa 100644
--- a/genesis-tools/genesis-populate/src/lib.rs
+++ b/genesis-tools/genesis-populate/src/lib.rs
@@ -29,7 +29,7 @@ use near_store::{
 use nearcore::{get_store_path, NightshadeRuntime};
 
 fn get_account_id(account_index: u64) -> AccountId {
-    AccountId::try_from(format!("near_{}_{}", account_index, account_index)).unwrap()
+    AccountId::try_from(format!("test{}", account_index)).unwrap()
 }
 
 pub type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;
diff --git a/runtime/runtime-params-estimator/src/testbed_runners.rs b/runtime/runtime-params-estimator/src/testbed_runners.rs
index 2ede73d9..ab316ae6 100644
--- a/runtime/runtime-params-estimator/src/testbed_runners.rs
+++ b/runtime/runtime-params-estimator/src/testbed_runners.rs
@@ -18,7 +18,7 @@ use std::time::Instant;
 
 /// Get account id from its index.
 pub fn get_account_id(account_index: usize) -> AccountId {
-    AccountId::try_from(format!("near_{}_{}", account_index, account_index)).unwrap()
+    AccountId::try_from(format!("test{}", account_index)).unwrap()
 }
 
 /// Total number of transactions that we need to prepare.
```

makes the test go fast. The inverse (changing all instances of `test{}` to `near_{}_{}` in the code base) plus adding random order of accounts makes the fast test slow.

**Comment by matklad** [comment]
Excellent find! I confirm that the `fast` test starts going slowly with `near_{}_{}` accounts, so this removes any param-estimator related weirdness from the equation. I also had to replace `test{}` with `near_{}_{}` across the codebase, and not just in a test, which felt odd. Filed https://github.com/near/nearcore/issues/4835

**Comment by bowenwang1996** [comment]
It feels that this is indeed related to the shape of the trie. @mina86 is it true that changing the names from `test{}` to `near_{}_{}` actually changes the number of database reads we do?

**Comment by mina86** [comment]
There’s only a small difference:

                                 test{}      near_{}_{}
    Trie::retrieve_raw_bytes   : 1832596  →  1944114
    TrieCachingStorage         : 2118923  →  2232747
    TrieCachingStorage (cached):       2  →        0
    TrieCachingStorage (missed): 2118921  →  2232747
    RocksDB::get               : 2322764  →  2436590

**Comment by bowenwang1996** [comment]
That is weird. Do we see any significant difference in the amount of time spent on `RocksDB::get`?

**Comment by matklad** [comment]
Modernized and simplified the repro based on mina86 refactors:

https://github.com/matklad/nearcore/commit/558b8e21142d14bcf9fd6fca16c03cabed6234b6


To run:

    cargo test --release --package runtime-tester --lib \
        --features nearcore/no_cache \
        -- perf_cliff --exact --nocapture

I've also found a better workflow for perf profiles

```
1. Build the test binary

    cargo test --no-run --release --package runtime-tester --features nearcore/no_cache --lib -- perf_cliff --exact --nocapture

2. Run the binary directly with perf (you can see the name of the file by ommitting --no-run). -F is sampling frequency: higher number, means bigger size of the resulting profile file

    perf record -F 999 --call-graph dwarf -- target/release/deps/runtime_tester-160c2544f9ae8411 perf_cliff --exact --nocapture

3. The output of the previous command is perf.data file. You need to process it with perf script (I have no idea what this actually does)

    perf script -F +pid > slow.perf

4. Go to https://profiler.firefox.com and drag and drop the slow.perf file
5. Click "Upload local Profile" in the top-right corner
6. Share the URL with colleagues
```

Here's what I get for slow (`near_{}_{}`) and fast (`test_{}`) versions:

* slow https://profiler.firefox.com/public/azmbf8q65nbrhdqj98752adjnkt18kj9c9pjpvr/flame-graph/?globalTrackOrder=0&localTrackOrderByPid=212964-0wxl&thread=1&v=6
* fast https://profiler.firefox.com/public/vv11t42jzk8nw9fagjd8s62j6g0q9h5s4d7jk6g/flame-graph/?globalTrackOrder=0&localTrackOrderByPid=213518-0wxo&thread=1&v=6

Note that profile is *per thread*, and we have a lot of threads. Select `perf_cliff` thread at the top and `Flame Graph` view at the bottom to get the most amusing result.

**Comment by bowenwang1996** [comment]
@matklad so it appears to be something inside RocksDB? cc @mina86

**Comment by matklad** [comment]
Aha, the stack issue is just becaues by default pref has a limit on stack depth. THe following argument fixes that: `--call-graph dwarf,65528`

Here's much better profiles

* slow (`near_{}_{}`) https://share.firefox.dev/2WJxwls
* fast (`test_{}`) https://share.firefox.dev/3mswaVh

**Comment by matklad** [comment]
Yeah, looking at those profiles gives so much more to work with. The difference seems to be that in the `rocksdb::Version::Get` the slow version does everything that the fast version does, *plus* `rocksdb::TableCache::Get`. I'll stop my investigation here and go back to wasm stuff.

**Comment by matklad** [comment]
This now I think blocks progress on https://github.com/near/nearcore/pull/4733. I am investigating many-orders-of-magnitude differences between the old estimator and the new one, and it seems that I hit "DB is mysteriously slow" case. 

The slow case is what is hit by the old estimator, and it looks like this: 

```
    1.78ms           Runtime::process_receipt
      1.78ms           Runtime::apply_action_receipt
        1.76ms           Runtime::apply_action
          1.76ms           Runtime::action_function_call
            1.76ms           Runtime::execute_function_call
              1.62ms           Runtime::get_code
                110.87µs         retrieve_raw_bytes
                104.19µs         retrieve_raw_bytes
                105.77µs         retrieve_raw_bytes
                105.73µs         retrieve_raw_bytes
                106.31µs         retrieve_raw_bytes
                104.89µs         retrieve_raw_bytes
                104.48µs         retrieve_raw_bytes
                104.49µs         retrieve_raw_bytes
                105.22µs         retrieve_raw_bytes
                114.13µs         retrieve_raw_bytes
                104.68µs         retrieve_raw_bytes
                110.49µs         retrieve_raw_bytes
                110.93µs         retrieve_raw_bytes
                110.50µs         retrieve_raw_bytes
                105.30µs         retrieve_raw_bytes
              138.43µs         Runtime::execute_function_call/1
                137.73µs         Runtime::execute_function_call/2
                  137.15µs         Runtime::execute_function_call/3
                    136.50µs         run_vm
```

It takes 1.5ms to fetch contract code from the database, which doesn't make any sense. The fast case looks like this: 


```
    210.09µs         Runtime::process_receipt
      209.44µs         Runtime::apply_action_receipt
        195.90µs         Runtime::apply_action
          195.07µs         Runtime::action_function_call
            192.66µs         Runtime::execute_function_call
              48.14µs          Runtime::get_code
                4.04µs           retrieve_raw_bytes
                1.48µs           retrieve_raw_bytes
                1.89µs           retrieve_raw_bytes
                1.87µs           retrieve_raw_bytes
                2.56µs           retrieve_raw_bytes
                1.83µs           retrieve_raw_bytes
                2.08µs           retrieve_raw_bytes
                2.90µs           retrieve_raw_bytes
                4.85µs           retrieve_raw_bytes
                3.55µs           retrieve_raw_bytes
                2.91µs           retrieve_raw_bytes
                2.01µs           retrieve_raw_bytes
                2.28µs           retrieve_raw_bytes
                2.93µs           retrieve_raw_bytes
                3.64µs           retrieve_raw_bytes
              143.66µs         Runtime::execute_function_call/1
                142.99µs         Runtime::execute_function_call/2
                  142.30µs         Runtime::execute_function_call/3
                    141.54µs         run_vm
```

**Comment by Longarithm** [comment]
What if rocksdb issues list is helpful here? This one is quite curious: https://github.com/facebook/rocksdb/issues/3961

When I disable `set_cache_index_and_filter_blocks` - https://github.com/near/nearcore/compare/558b8e211..bceefb56 - performance become much more stable:
- disabled: 40-80ms to process a single block in most cases, one spike with 120ms
- enabled: 40ms, then **1s for each block**

@ailisp do you know why we enable it?

**Comment by pmnoxx** [comment]
@Longarithm We can try to see if it makes difference to try to adjust cache sizes per column. Like in https://github.com/near/nearcore/pull/5212

**Comment by pmnoxx** [comment]
> have a lot of threads. Select `perf_cliff` thread at the top and `Flame Graph` view at the bottom to get the most amusing result.

It appears that int the slow path we spend a lot of time int `rocksdb::BlockFetcher::ReadBlockContents`. So RocksDB is spending a lot of time reading blocks in the slow case.

**Comment by pmnoxx** [comment]
> retrieve_raw_bytes

I see that we have a cache in front of the database. This explains the spikes in performance. If you want to debug the issue, please try commenting this code out:

``
impl TrieStorage for TrieCachingStorage {
    fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Result<Vec<u8>, StorageError> {
        let mut guard = self.cache.0.lock().expect(POISONED_LOCK_ERR);
        /*    if let Some(val) = guard.cache_get(hash) {
            Ok(val.clone())
        } else {*/
        let key = Self::get_key_from_shard_uid_and_hash(self.shard_uid, hash);
        let val = self
            .store
            .get(ColState, key.as_ref())
            .map_err(|_| StorageError::StorageInternalError)?;
        if let Some(val) = val {
            if val.len() < TRIE_LIMIT_CACHED_VALUE_SIZE {
                guard.cache_set(*hash, val.clone());
            }
            Ok(val)
        } else {
            // not StorageError::TrieNodeMissing because it's only for TrieMemoryPartialStorage
            Err(StorageError::StorageInconsistentState("Trie node missing".to_string()))
        }
        //     }
    }

    fn as_caching_storage(&self) -> Option<&TrieCachingStorage> {
        Some(self)
    }
}
```


This makes reads slower, but this shows you the real time it takes to access the database.
```
  462.38ms         Process block
    461.43ms         process_state_update
      461.42ms         Runtime::apply
        66.14ms   100    Runtime::process_receipt
          100.10µs  100    get_key
        206.69ms  100    Runtime::process_transaction

  458.03ms         Process block
    457.15ms         process_state_update
      457.14ms         Runtime::apply
        67.32ms   100    Runtime::process_receipt
          94.68µs   100    get_key
        204.97ms  100    Runtime::process_transaction

  508.51ms         Process block
    507.50ms         process_state_update
      507.49ms         Runtime::apply
        68.10ms   100    Runtime::process_receipt
          97.73µs   100    get_key
        204.74ms  100    Runtime::process_transaction

  469.07ms         Process block
    468.15ms         process_state_update
      468.14ms         Runtime::apply
        68.92ms   100    Runtime::process_receipt
          100.43µs  100    get_key
        211.39ms  100    Runtime::process_transaction

  459.90ms         Process block
    459.00ms         process_state_update
      458.99ms         Runtime::apply
        67.22ms   100    Runtime::process_receipt
          95.58µs   100    get_key
        205.08ms  100    Runtime::process_transaction

  464.87ms         Process block
    463.96ms         process_state_update
      463.96ms         Runtime::apply
        68.66ms   100    Runtime::process_receipt
          97.19µs   100    get_key
        207.97ms  100    Runtime::process_transaction

```

**Comment by pmnoxx** [comment]
I created a benchmark https://github.com/near/nearcore/pull/5312/files

It may be useful to try different database configuration.

**Comment by pmnoxx** [comment]
The database is becoming slower and slower as number of keys grows
```
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

     Running unittests (/home/pmnoxx/repos/nearcore.master/target/release/deps/ibf-b2f9e4ff4ef48925)

running 1 test
test benchmark_db ... took on avg 7.957µs
took on avg 7.484µs
took on avg 8.615µs
took on avg 9.071µs
took on avg 8.883µs
took on avg 9.11µs
took on avg 7.115µs
took on avg 7.675µs
took on avg 8.498µs
took on avg 7.6µs
took on avg 7.692µs
took on avg 8.218µs
took on avg 8.48µs
took on avg 9.615µs
^[[Atook on avg 9.806µs
took on avg 9.733µs
took on avg 10.69µs
took on avg 10.832µs
^[took on avg 10.905µs
took on avg 12.207µs
took on avg 12.522µs
took on avg 12.604µs
took on avg 12.835µs
took on avg 13.551µs
took on avg 14.082µs
took on avg 14.72µs
took on avg 15.908µs
took on avg 15.729µs
took on avg 16.746µs
took on avg 17.339µs
took on avg 18.038µs
took on avg 18.902µs
took on avg 17.711µs
took on avg 19.251µs
took on avg 20.995µs
```

**Comment by pmnoxx** [comment]
@bowenwang1996 @mm-near  @matklad @mina86 You can play with some database setting to try to make it better.
all are inside `db.rs`.

**Comment by pmnoxx** [comment]
I added extra stats
```
took on avg 1.433µs op per sec 697554 
took on avg 1.524µs op per sec 655795 
took on avg 2.017µs op per sec 495621 
took on avg 1.577µs op per sec 633970 
took on avg 2.044µs op per sec 489095 
took on avg 2.5µs op per sec 399938 
took on avg 2.092µs op per sec 477793 
took on avg 3.054µs op per sec 327352 
took on avg 2.103µs op per sec 475298 
took on avg 2.105µs op per sec 474993 
took on avg 2.66µs op per sec 375821 
took on avg 2.478µs op per sec 403493 
took on avg 2.733µs op per sec 365783 
took on avg 3.094µs op per sec 323138 
took on avg 2.696µs op per sec 370820 
took on avg 3.757µs op per sec 266117 
took on avg 3.264µs op per sec 306341 
took on avg 3.703µs op per sec 270033 
took on avg 3.153µs op per sec 317110 
took on avg 2.712µs op per sec 368696 
took on avg 3.727µs op per sec 268249 

```

With `3.727µs` we are doing 268249 ops/s. That's considered quite good for a database. 
Stats from https://github.com/facebook/rocksdb/wiki/Performance-Benchmarks#test-1-bulk-load-of-keys-in-random-order stop results with half of that value. Those our production machines are probably 2-3 slower than my desktop.

Ssds are able to achieve 300k IOPS (https://www.enterprisestorageforum.com/hardware/how-fast-are-nvme-speeds/)
So there is no problem there.

**Comment by pmnoxx** [comment]
Here are some guides on optimizing rocks db: https://github-wiki-see.page/m/facebook/rocksdb/wiki/RocksDB-Tuning-Guide

Btw,
```
    opts.set_max_open_files(512);
```
Seems quite low. We have a lot of open columns, closing files and opening them could cause performance issues.

It may be a good idea to check number of open files on live node.

**Comment by pmnoxx** [comment]
It's also likely that something like compaction not being configured right, may be causing huge latency spikes, likes one I saw in benchmark.

**Comment by bowenwang1996** [comment]
@Longarithm please take a look at @pmnoxx's benchmark

**Comment by pmnoxx** [comment]
@bowenwang1996 @Longarithm 

In the test I write 100k or 1m or 10m keys to the database, but I'm only able to read 48.5% of them back successfully. Something seems broken.

```
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

     Running unittests (/home/pmnoxx/repos/nearcore.master/target/release/deps/ibf-0ff475f5b700c766)

running 1 test
test benchmark_db ... 

took on avg 2.646µs op per sec 377805 got 4879485/10000000

```

**Comment by Longarithm** [comment]
I hit this problem in another situation.

I wanted to iterate over Trie contracts, because of the structure the iterator at first went over all accounts, and then over all contracts. Around 15 Dec, it finished in ~1 minute, but now it takes extremely long time. In the example below, it takes ~2 seconds to iterate over 100 accounts, which is not reasonable:

```
Jan 12 14:26:32.545  INFO neard: 84100
Jan 12 14:26:34.481  INFO neard: 84200
Jan 12 14:26:36.506  INFO neard: 84300
Jan 12 14:26:38.448  INFO neard: 84400
Jan 12 14:26:40.233  INFO neard: 84500
Jan 12 14:26:42.175  INFO neard: 84600
Jan 12 14:26:44.122  INFO neard: 84700
Jan 12 14:26:45.462  INFO neard: 84800
Jan 12 14:26:47.045  INFO neard: 84900
Jan 12 14:26:49.073  INFO neard: 85000
```

https://github.com/near/nearcore/compare/e7e9fbc30e9de...c0078eab22 - this code stops on `84000`th account, but it doesn't matter.

Also I believe at *some* moment the code became fast again, and also if I iterate over N accounts in some **previous** run, then in the next run these accounts are traversed much faster. This makes the issue much harder to debug, and makes me think that some rocksdb caching steps in here. Profiler shows that `rocksdb::BlockBasedTable::Get` is the bottleneck: https://share.firefox.dev/3K6VIC8

Though it can be avoided for my particular task, I feel more worried about db performance now.

**Comment by jakmeier** [comment]
I am investigating a similar issue to this one. I think I might have found the explanation: In the slow case, RocksDB is reading state from disk and in the fast case it's all in memory.

## My test scenario 

In my tests, I create 6000 new accounts with names `format!("{hash(i)}_near")`, giving much more randomly distributed accounts within the trie structure than any of the tests above.

Reading 400 of those accounts right after their creation from the estimator, I can usually get an account from RocksDB within ~100us. This is remarkably fast because the estimator avoids shard and chunk caches and it clears RocksDB write-back buffers. I also reduced RocksDB block cache to1 kiB but the access is still fast.
With all those caches disabled, the expectation is that reading an account will trigger a trie traversal with every node being read from disk. But 100us is at most one SSD access, more likely it is in-memory only.
I checked that requests to the DB are sent as expected and it seems to be the case. But I noticed that `State` column access times of merely ~10us. Looks like RocksDB can serve these requests from memory.

So then if I drop OS caches between creating the accounts and reading them, and all of a sudden the `State` column access time goes up to 7-8ms. Retrieving an account takes at least one such access but could be up to ~20 for trie traversal. With shard cache enabled, it's usually just 2 new nodes per account (extension + leaf), which gives ~15ms per account. This is perfectly in line with 1-2 seconds to iterate over 100 accounts as has been mentioned in older comments.

## My conclusion 

So, in my case I can trigger disk lookups by dropping the OS page cache. And I see the performance degradation equal to that described in this issue. It could be that the scenarios analysed in this issue also hit the OS page cache.

The Linux page cache is interesting in that it takes all the available unused RAM, which can be many GBs or it can be close to non-existent if the machine's memory is used otherwise. 

My guess is that the performance cliff is due to a complicated combination of available page cache size, how accounts are distributed over SST files and in which layer. (Remember that SST files in level 0 are not sorted while all other are. So in layer 0, insertion order matters more than global key order.) Depending on how many SST files need to be fetched, how much fits in the RocksDB block cache, and how much of it fits in the Page Cache, there will be almost no disk accesses (fast case) or there is a new SST file read from disk every dozen or so accounts (slow case) because those accounts are in different SST files in level 0.

**Comment by EdvardD** [comment]
> The rocksdb::VerifyBlockChecksum looks very suspicious!

I'm actually already working on that. See the PR with the fork-fix: https://github.com/near/nearcore/pull/7418 and fix in the rocksdb: https://github.com/facebook/rocksdb/pull/10427/files

**Comment by jakmeier** [comment]
Applying #7418 changes does not resolve my testcase. It's still taking milliseconds to read single values. There seems to be good improvement but not the 1000x we are looking for. I guess this it not too surprising, as my test is only slow if I reset OS page caches, which clearly points to this being disk bound and not CPU bound.

But also, my test case is no longer perfectly consistent anymore, which it was the previous two days. So maybe I was a bit too confident in my previous claims...

**Comment by jakmeier** [comment]
**TLDR** I found out quite a bit more by now. The slow case observed seems to be mostly due to a bad SST file layout. Whereas the fast case was all served from in-memory write-back buffers. The truth is somewhere in between, probably.
___ 

The first big discovery was that instead of clearing OS page caches, which took usually about 2 seconds, I could also do `thread::sleep(Duration::from_millis(2000))` to get the same result. Without pause, I get the fast case. With pause, I get almost 1000x slower performance.

Then I managed to get a case where a loop starts with a couple fast iterations and then goes over to slow DB reads. At this very moment, a RocksDB flush of the `State` column family kicked in. This happens in the background if I wait for long enough but it seems it can be delayed if we are putting enough load on RocksDB otherwise.

The problem is, there is already code to specifically force a flush before each measurement to avoid this problem. Turns out, the forced flush I added some time ago is buggy. See #7484 for the fix. With that fix in, things get really slow. Qemu based estimations showed a steep increase in read bytes and time based estimations sometimes go up by 10x. Seems that the flushing of memtables never really worked as intended.

Now the problem was that things are _way_ too slow. With a 40GB database setup on my local SSD, I managed to get estimations >1s per empty action receipt. About 10'000 times slower than before. Or when measured with qemu, more than 1GB was being read per empty action receipt! With smaller DBs, everything is fast and small read throughput.

The final puzzle peace is that RocksDB will produce quite unfortunate SST file layouts when it is forced to quickly store a lot of data and doesn't get time to clean up. The estimator creates a new DB from a state dump everytime. In other words, all data is written at once right before starting the estimation. Now with the additional force-flush in between. Exactly the recipe for bad SST file layout.

The solution it to run compaction on the newly created DB. With that simply change, I can go back and forth between >1s and 650us per empty action receipt. This is somewhere between the slowest (1s+) and fastest (~100us) cases I could reliably reproduce.

Note that all numbers so far are without dropping OS page cache between measurement. It is not required to explain the huge differences reported in this github issue. But when I do drop the page cache with all fixes applied, the measurement goes from 650us up to 3ms, so maybe 5x times slower. This is a completely reasonable time for a workload that has to go to disk dozens of times.

Now that all of this is explained and written down, I will close this issue. I will post more updates in #5995 with respect to necessary changes to the estimator.

**Comment by matklad** [comment]
>I managed to get estimations >1s per empty action receipt. About 10'000 times slower than before. 

There should be a Russian-reversal joke how at NEAR we are working hard on making the slowest possible code! 

Kudos for getting to the bottom of this and not letting this bug to celebrate its first birthday!
